File: src/runtimeCatalog.ts
import { HealthRegistry } from "./health.js";
import { AppConfig, ModelCatalog, ModelConfig, RuntimeModelMetadata, SelectedModel } from "./types.js";

export class RuntimeCatalog implements ModelCatalog {
  private readonly overlays = new Map<string, RuntimeModelMetadata>();
  private readonly virtualModels = new Map<string, ModelConfig>();

  constructor(
    private readonly config: AppConfig,
    private readonly health: HealthRegistry,
  ) {}

  setOverlay(modelId: string, metadata: RuntimeModelMetadata) {
    this.overlays.set(modelId, metadata);
  }

  setVirtualModel(modelId: string, model: ModelConfig, metadata: RuntimeModelMetadata) {
    this.virtualModels.set(modelId, model);
    this.overlays.set(modelId, metadata);
    this.health.ensureModel(modelId, model);
  }

  getModel(modelName: string): SelectedModel | undefined {
    const configured = this.config.models[modelName];
    if (configured) return { name: modelName, config: effectiveModel(configured, this.overlays.get(modelName)), metadata: this.overlays.get(modelName) };
    const virtual = this.virtualModels.get(modelName);
    if (virtual) return { name: modelName, config: effectiveModel(virtual, this.overlays.get(modelName)), metadata: this.overlays.get(modelName) };
    return undefined;
  }

  getModelEntries(): Array<[string, ModelConfig, RuntimeModelMetadata | undefined]> {
    return [
      ...Object.entries(this.config.models).map(([id, model]) => [id, effectiveModel(model, this.overlays.get(id)), this.overlays.get(id)] as [string, ModelConfig, RuntimeModelMetadata | undefined]),
      ...Array.from(this.virtualModels.entries()).map(([id, model]) => [id, effectiveModel(model, this.overlays.get(id)), this.overlays.get(id)] as [string, ModelConfig, RuntimeModelMetadata | undefined]),
    ];
  }

  getConfiguredModelEntries(): Array<[string, ModelConfig]> {
    return Object.entries(this.config.models);
  }

  getVirtualModelIds(): string[] {
    return Array.from(this.virtualModels.keys()).sort();
  }

  getRuntimeMetadataSummary(): Record<string, RuntimeModelMetadata> {
    return Object.fromEntries(Array.from(this.overlays.entries()).sort(([a], [b]) => a.localeCompare(b)));
  }
}

function effectiveModel(model: ModelConfig, metadata?: RuntimeModelMetadata): ModelConfig {
  if (!metadata) return model;
  return {
    ...model,
    context_window: metadata.context_window ?? model.context_window,
    cost_tier: metadata.dynamic_cost_tier ?? model.cost_tier,
    price_rank: metadata.dynamic_price_rank ?? model.price_rank,
    supported_parameters: metadata.supported_parameters ?? model.supported_parameters,
  };
}

File: rust/runtime_catalog.rs
use std::{
    collections::HashMap,
    sync::{Arc, RwLock},
};

use crate::{
    health::HealthRegistry,
    types::{AppConfig, ModelConfig, RuntimeModelMetadata, SelectedModel},
};

#[derive(Clone)]
pub struct RuntimeCatalog {
    config: Arc<AppConfig>,
    health: HealthRegistry,
    overlays: Arc<RwLock<HashMap<String, RuntimeModelMetadata>>>,
    virtual_models: Arc<RwLock<HashMap<String, ModelConfig>>>,
}

impl RuntimeCatalog {
    pub fn new(config: Arc<AppConfig>, health: HealthRegistry) -> Self {
        Self {
            config,
            health,
            overlays: Arc::new(RwLock::new(HashMap::new())),
            virtual_models: Arc::new(RwLock::new(HashMap::new())),
        }
    }

    pub fn set_overlay(&self, model_id: &str, metadata: RuntimeModelMetadata) {
        self.overlays
            .write()
            .expect("catalog overlay lock poisoned")
            .insert(model_id.to_string(), metadata);
    }

    pub fn set_virtual_model(
        &self,
        model_id: &str,
        model: ModelConfig,
        metadata: RuntimeModelMetadata,
    ) {
        self.virtual_models
            .write()
            .expect("catalog virtual lock poisoned")
            .insert(model_id.to_string(), model.clone());
        self.set_overlay(model_id, metadata);
        self.health.ensure_model(model_id, Some(model), false);
    }

    pub fn get_model(&self, model_name: &str) -> Option<SelectedModel> {
        let overlays = self.overlays.read().expect("catalog overlay lock poisoned");
        if let Some(configured) = self.config.models.get(model_name) {
            let metadata = overlays.get(model_name).cloned();
            return Some(SelectedModel {
                name: model_name.to_string(),
                config: effective_model(configured, metadata.as_ref()),
                metadata,
            });
        }
        let virtual_model = self
            .virtual_models
            .read()
            .expect("catalog virtual lock poisoned")
            .get(model_name)
            .cloned();
        virtual_model.map(|model| {
            let metadata = overlays.get(model_name).cloned();
            SelectedModel {
                name: model_name.to_string(),
                config: effective_model(&model, metadata.as_ref()),
                metadata,
            }
        })
    }

    pub fn get_model_entries(&self) -> Vec<(String, ModelConfig, Option<RuntimeModelMetadata>)> {
        let overlays = self.overlays.read().expect("catalog overlay lock poisoned");
        let mut entries: Vec<_> = self
            .config
            .models
            .iter()
            .map(|(id, model)| {
                let metadata = overlays.get(id).cloned();
                (
                    id.clone(),
                    effective_model(model, metadata.as_ref()),
                    metadata,
                )
            })
            .collect();
        entries.extend(
            self.virtual_models
                .read()
                .expect("catalog virtual lock poisoned")
                .iter()
                .map(|(id, model)| {
                    let metadata = overlays.get(id).cloned();
                    (
                        id.clone(),
                        effective_model(model, metadata.as_ref()),
                        metadata,
                    )
                }),
        );
        entries
    }

    pub fn get_configured_model_entries(&self) -> Vec<(String, ModelConfig)> {
        self.config
            .models
            .iter()
            .map(|(id, model)| (id.clone(), model.clone()))
            .collect()
    }

    pub fn get_virtual_model_ids(&self) -> Vec<String> {
        let mut ids: Vec<_> = self
            .virtual_models
            .read()
            .expect("catalog virtual lock poisoned")
            .keys()
            .cloned()
            .collect();
        ids.sort();
        ids
    }

    pub fn get_runtime_metadata_summary(&self) -> HashMap<String, RuntimeModelMetadata> {
        self.overlays
            .read()
            .expect("catalog overlay lock poisoned")
            .clone()
    }
}

pub fn effective_model(
    model: &ModelConfig,
    metadata: Option<&RuntimeModelMetadata>,
) -> ModelConfig {
    let Some(metadata) = metadata else {
        return model.clone();
    };
    let mut model = model.clone();
    if metadata.context_window.is_some() {
        model.context_window = metadata.context_window;
    }
    if metadata.dynamic_cost_tier.is_some() {
        model.cost_tier = metadata.dynamic_cost_tier.clone();
    }
    if metadata.dynamic_price_rank.is_some() {
        model.price_rank = metadata.dynamic_price_rank;
    }
    if metadata.supported_parameters.is_some() {
        model.supported_parameters = metadata.supported_parameters.clone();
    }
    model
}
